diff --git a/cli/tsc/dts/lib.deno.unstable.d.ts b/cli/tsc/dts/lib.deno.unstable.d.ts
index c87ac9ed8..2db1c9890 100644
--- a/cli/tsc/dts/lib.deno.unstable.d.ts
+++ b/cli/tsc/dts/lib.deno.unstable.d.ts
@@ -205,12 +205,13 @@ declare namespace Deno {
    *
    *  The parameters correspond to the table below:
    *
-   *  | system            | winHandle     | displayHandle   |
-   *  | ----------------- | ------------- | --------------- |
-   *  | "cocoa" (macOS)   | -             | `NSView*`       |
-   *  | "win32" (Windows) | `HWND`        | `HINSTANCE`     |
-   *  | "x11" (Linux)     | Xlib `Window` | Xlib `Display*` |
-   *  | "wayland" (Linux) | `wl_surface*` | `wl_display*`   |
+   *  | system                     | winHandle       | displayHandle   |
+   *  | -------------------------- | --------------- | --------------- |
+   *  | "cocoa" (macOS)            | -               | `NSView*`       |
+   *  | "core-animation" (macOS)   | `CAMetalLayer*` | -               |
+   *  | "win32" (Windows)          | `HWND`          | `HINSTANCE`     |
+   *  | "x11" (Linux)              | Xlib `Window`   | Xlib `Display*` |
+   *  | "wayland" (Linux)          | `wl_surface*`   | `wl_display*`   |
    *
    * @category GPU
    * @experimental
@@ -223,7 +224,7 @@ declare namespace Deno {

     constructor(
       options: {
-        system: "cocoa" | "win32" | "x11" | "wayland";
+        system: "cocoa" | "core-animation" | "win32" | "x11" | "wayland";
         windowHandle: Deno.PointerValue<unknown>;
         displayHandle: Deno.PointerValue<unknown>;
         width: number;
diff --git a/ext/canvas/byow.rs b/ext/canvas/byow.rs
index 18ab2800b..bdcaf41f4 100644
--- a/ext/canvas/byow.rs
+++ b/ext/canvas/byow.rs
@@ -185,6 +185,29 @@ impl UnsafeWindowSurface {
       return Err(ByowError::InvalidParameters);
     }

+    #[cfg(target_os = "macos")]
+    if options.system == UnsafeWindowSurfaceSystem::CoreAnimation {
+      // SAFETY: see above comment
+      let id = unsafe {
+        instance
+          .instance_create_surface_metal(
+            options.window_handle as *mut c_void,
+            None,
+          )
+          .map_err(ByowError::CreateSurface)?
+      };
+
+      return Ok(UnsafeWindowSurface {
+        data: Rc::new(RefCell::new(SurfaceData {
+          width: options.width,
+          height: options.height,
+          id,
+          instance,
+        })),
+        active_context: Default::default(),
+      });
+    }
+
     let (win_handle, display_handle) = raw_window(
       options.system,
       options.window_handle,
@@ -311,6 +334,7 @@ struct UnsafeWindowSurfaceOptions {
 #[derive(Eq, PartialEq)]
 enum UnsafeWindowSurfaceSystem {
   Cocoa,
+  CoreAnimation,
   Win32,
   X11,
   Wayland,
@@ -334,6 +358,7 @@ impl<'a> FromV8<'a> for UnsafeWindowSurfaceOptions {
     let s = String::from_v8(scope, val).unwrap();
     let system = match s.as_str() {
       "cocoa" => UnsafeWindowSurfaceSystem::Cocoa,
+      "core-animation" => UnsafeWindowSurfaceSystem::CoreAnimation,
       "win32" => UnsafeWindowSurfaceSystem::Win32,
       "x11" => UnsafeWindowSurfaceSystem::X11,
       "wayland" => UnsafeWindowSurfaceSystem::Wayland,
